sdk/identity/src/token_credentials/workload_identity_credentials.rs

```diff
@@ -17,37 +23,87 @@ use time::OffsetDateTime;
 #[derive(Debug)]
 pub struct WorkloadIdentityCredential {
     http_client: Arc<dyn HttpClient>,
+    authority_host: Url,
     tenant_id: String,
     client_id: String,
     token: Secret,
-    options: TokenCredentialOptions,
     cache: TokenCache,
 }
 
 impl WorkloadIdentityCredential {
     /// Create a new `WorkloadIdentityCredential`
     pub fn new<T>(
-        http_client: Arc<dyn HttpClient>,
+        options: impl Into<TokenCredentialOptions>,
         tenant_id: String,
         client_id: String,
         token: T,
     ) -> Self
     where
         T: Into<Secret>,
     {
+        let options = options.into();
         Self {
-            http_client,
+            http_client: options.http_client().clone(),
+            authority_host: options.authority_host().clone(),
             tenant_id,
             client_id,
             token: token.into(),
-            options: TokenCredentialOptions::default(),
             cache: TokenCache::new(),
         }
     }
 
-    /// set `TokenCredentialOptions`
-    pub fn set_options(&mut self, options: TokenCredentialOptions) {
-        self.options = options;
+    pub fn create(
+        options: impl Into<TokenCredentialOptions>,
+    ) -> azure_core::Result<WorkloadIdentityCredential> {
+        let options = options.into();
+        let env = options.env();
+        let tenant_id =
+            env.var(AZURE_TENANT_ID_ENV_KEY)
+                .with_context(ErrorKind::Credential, || {
+                    format!(
+                        "working identity credential requires {} environment variable",
+                        AZURE_TENANT_ID_ENV_KEY
+                    )
+                })?;
```

Comment:
```suggestion
                .map_kind(ErrorKind::Credential)?;
```

Reply:
The error messages are indeed better with how it is. I provided an example [above](https://github.com/Azure/azure-sdk-for-rust/pull/1532#discussion_r1439922931).